In [1]:
pip install -U dspy


Note: you may need to restart the kernel to use updated packages.



[notice] A new release of pip is available: 25.1.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [2]:
import dspy

In [3]:
print("DSPy version:", dspy.__version__)

DSPy version: 3.0.3


In [4]:
lm=dspy.LM("gemini/gemini-2.5-flash", api_key="AIzaSyBGpmviDEBJiPpi44vyZOHvO7Zen-1Cmus")
dspy.configure(lm=lm)

In [5]:
math = dspy.ChainOfThought("question -> answer: float")
math(question="Two dice are tossed. What is the probability that the sum equals two?")

Prediction(
    reasoning='To find the probability, we need to determine the number of favorable outcomes and divide it by the total number of possible outcomes.\n\n1.  **Total Possible Outcomes:**\n    When two dice are tossed, each die has 6 faces (1, 2, 3, 4, 5, 6).\n    The total number of possible outcomes is 6 * 6 = 36.\n    These outcomes can be represented as pairs (die1, die2):\n    (1,1), (1,2), (1,3), (1,4), (1,5), (1,6)\n    (2,1), (2,2), (2,3), (2,4), (2,5), (2,6)\n    (3,1), (3,2), (3,3), (3,4), (3,5), (3,6)\n    (4,1), (4,2), (4,3), (4,4), (4,5), (4,6)\n    (5,1), (5,2), (5,3), (5,4), (5,5), (5,6)\n    (6,1), (6,2), (6,3), (6,4), (6,5), (6,6)\n\n2.  **Favorable Outcomes (Sum equals two):**\n    We need to find the combinations where the sum of the two dice is 2.\n    The only combination that results in a sum of 2 is (1, 1).\n    So, there is only 1 favorable outcome.\n\n3.  **Calculate Probability:**\n    Probability = (Number of Favorable Outcomes) / (Total Possible Ou

# basic predictions

In [6]:
predict=dspy.Predict("question -> answer")
response= predict(question= "what is the capital city of france")
print(f"answer: {response.answer}")

answer: Paris


# signature

In [7]:
class QnaSignature(dspy.Signature):
    """" Answer the question based on the context provided"""
    question= dspy.InputField(desc="the question to be answered.")
    context= dspy.InputField(desc="the context to answer the question.")
    answer=dspy.OutputField(desc="the answer to the question")
    
#create a module that uses the Signature
Predicator= dspy.Predict(QnaSignature)
result= Predicator(question= "which team won ipl 2025 ?",
                  context= "in june 2025, ipl final was held at naredra modi stadium , punjab and rcb played final match and finally rcb won the ipl after 18 years and its 18th year  ")

print("response:",result.answer)
    

response: RCB won the IPL 2025.


# Chain Of Thaught

In [8]:
class MathSig(dspy.Signature):
    """"solve the problem , include step by step reasoning then final answer """
    problem= dspy.InputField(desc="math problem as text")
    resoning=dspy.OutputField(desc="step by step chain of thaught")
    finalAnswer=dspy.OutputField(desc="conceise the final answer")
    
solver=dspy.ChainOfThought(MathSig)

res=solver(problem="if 3 boxes contain 4,5,6 apples, how many applles are there in total? " )
print("Reasoning: \n", getattr(res, "reasoning",res.get("reasoning")))

print("final answer:", getattr(res, "final answer", res.get("finalAnswer")))

Reasoning: 
 The problem asks for the total number of apples in three boxes. To find the total, I need to add the number of apples in each box together. The number of apples in the boxes are 4, 5, and 6.
final answer: 15


In [9]:
class MathReasoning(dspy.Signature):
    """"solve the problem , include step by step reasoning then final answer """
    problem= dspy.InputField(desc="math word problem ")
    resoning=dspy.OutputField(desc="step by step Reasoning")
    finalAnswer=dspy.OutputField(desc="only numeric number ")
    
solver=dspy.ChainOfThought(MathSig)

res=solver(problem="a shopkeeper buys 12 pens each costing 5 rupees , how much the total cost ? " )
print("Reasoning: \n", res.reasoning)

print("final answer:", res.finalAnswer)

Reasoning: 
 The problem asks for the total cost of 12 pens, with each pen costing 5 rupees. To find the total cost, I need to multiply the number of pens by the cost of one pen.
final answer: The total cost is 60 rupees.


# multi hop question answering

In [10]:
class MultiHopQA(dspy.Signature):
    question=dspy.InputField()
    context=dspy.InputField()
    reasoning=dspy.OutputField()
    finalAnswer=dspy.OutputField()
    
qa= dspy.ChainOfThought(MultiHopQA)
context= """" Narendra Damodardas Modi[a] (born 17 September 1950) is an Indian politician who has served as the prime minister of India since 2014. Modi was the chief minister of Gujarat from 2001 to 2014 and is the member of parliament (MP) for Varanasi. He is a member of the Bharatiya Janata Party (BJP) and of the Rashtriya Swayamsevak Sangh (RSS), a right-wing Hindutva paramilitary volunteer organisation. He is the longest-serving prime minister outside the Indian National Congress.

Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat), where he completed his secondary education. He was introduced to the RSS at the age of eight, becoming a full-time worker for the organisation in Gujarat in 1971. The RSS assigned him to the BJP in 1985, and he rose through the party hierarchy, becoming general secretary in 1998.[b] In 2001"""

res=qa(question="where did naredra modi born and when is his birthday?",context=context)

print("reasoning:\n", res.reasoning)
print("finalanswer:", res.finalAnswer)

reasoning:
 The question asks for Narendra Modi's birthplace and birthday. The context states that he was "born 17 September 1950" and "Modi was born and raised in Vadnagar, Bombay State (present-day Gujarat)". I will extract these two pieces of information.
finalanswer: Narendra Modi was born in Vadnagar, Bombay State (present-day Gujarat), and his birthday is on 17 September 1950.


# RAG+Chain of thaught

In [11]:
class RagQA(dspy.Signature):
    question=dspy.InputField()
    context=dspy.InputField()
    reasoning=dspy.OutputField()
    finalAnswer=dspy.OutputField()
    
rag_qa = dspy.ChainOfThought(RagQA)
docs="""peration Sindoor refers to a recent operation where the Indian armed forces targeted terrorist infrastructure in Pakistan and Pakistan-occupied Jammu and Kashmir, leading to Pakistan's request for a cessation of firing on May 10, 2025. Documentaries and official reports detail the tactical success, the role of indigenous defense technology, and India's strategic response, including countering misinformation and highlighting the commitment to dismantle terrorism. 
Key Aspects of Operation Sindoor
Target:
The operation targeted terrorist infrastructure in Pakistan and Pakistan-occupied Jammu and Kashmir. 
Timeline:
India launched Operation Sindoor in the early hours of May 7, 2025. 
Pakistan's Request:
Pakistan's Director General of Military Operations requested a cessation of firing on May 10, 2025, which was agreed to the same day. 
Strategic Objective:
The politico-military objective was to punish Pakistan for provocations and to punish terrorism at its roots, according to the Ministry of External Affairs and the Prime Minister. 
Documents and Media Coverage
Documentary: INSIDE OPERATION SINDOOR–The 88 Hours That Defined India by CNN-News18 features former military officials and analysts discussing the operation. 
Official Press Releases: The Press Information Bureau (PIB) and Ministry of External Affairs (MEA) have published reports on the operation and its implications. 
News Reports: News18 and YouTube offer detailed coverage of the event and its context. 
Significance
Indigenous Technology:
Operation Sindoor highlighted the effectiveness of India's indigenous defense technologies, from air defense systems to drones and net-centric warfare platforms. 
Information Warfare:
India countered a Pakistani misinformation campaign by providing facts, transparency, and digital vigilance. 
Assertive Stance:
The operation solidified India's image as an assertive and resolute nation committed to combating terrorism and defending its territory."""

res=rag_qa(question ="what is the cause of operation sindoor",context=docs)

print("reasoning:\n ", res.reasoning)

print("finalanswer:", res.finalAnswer)


reasoning:
  The context explicitly states the strategic objective of Operation Sindoor. It mentions that the politico-military objective was "to punish Pakistan for provocations and to punish terrorism at its roots." This directly answers the question about the cause of the operation.
finalanswer: Operation Sindoor was launched to punish Pakistan for provocations and to punish terrorism at its roots.


# REACT = REASON+ACT

In [12]:
#dummy retriver
class Retriever:
    def search(self,query,k=2):
        return [f"doc about {query} with facts."]
    
class QnASig(dspy.Signature):
    question=dspy.InputField()
    answer=dspy.OutputField()
retriever = Retriever()

qa_agent=dspy.ReAct(QnASig, tools=[retriever.search])

res=qa_agent(question="what is Machine Learning?")
print(res.answer)

Machine Learning is a subfield of Artificial Intelligence (AI) that focuses on enabling systems to learn from data, identify patterns, and make decisions or predictions with minimal human intervention. Instead of being explicitly programmed for every task, machine learning algorithms build a model from sample data (known as "training data") to automate analytical model building. This allows computers to find hidden insights without being explicitly programmed where to look.


# E-commerce Assistant

In [13]:
class ProductSig(dspy.Signature):
    user_request=dspy.InputField()
    response=dspy.OutputField()

def check_inventory(item: str):
    stock = {"iphone": 5, "samsung": 0}
    # normalize item to lowercase so lookups match
    item_key = item.lower()
    return f"{item_key} stock : {stock.get(item_key, 'unknown')}"

# Wrap the function as a dspy Tool so the agent can call it
check_tool = dspy.Tool(
    func=check_inventory,
    name="check_inventory",
    desc="Return the current stock for the given product. Arg name: item (string)."
)

# create the agent with the Tool (list of Tool objects)
shop_bot = dspy.ReAct(ProductSig, tools=[check_tool])

# query the agent
res = shop_bot(user_request="do you have iphone phones in stock?")

print(res.response)

Yes, we have 5 iPhones in stock.


# program of thought

In [14]:
pip install deno



[notice] A new release of pip is available: 25.1.1 -> 25.2
[notice] To update, run: python.exe -m pip install --upgrade pip


In [15]:
class MathsSig(dspy.Signature):
    question=dspy.InputField()
    answer=dspy.OutputField()
    
math_solver=dspy.ProgramOfThought(MathsSig)

res=math_solver(question="what is (25*4)+(60/3)?")

print(res.answer)

120


# data cleaning assistant-program of thought

In [16]:
class CleanSig(dspy.Signature):
    instruction=dspy.InputField()
    cleaned_data=dspy.OutputField()
data_cleaner=dspy.ProgramOfThought(CleanSig)
res=data_cleaner(instruction="remove duplicates from [1,2,2,3,3,4]")

print(res.cleaned_data)

[1, 2, 3, 4]


# invoice validation-program of thought

In [17]:
class invoiceSig(dspy.Signature):
    invoice = dspy.InputField()
    total = dspy.OutputField()
invoice_solver = dspy.ProgramOfThought(invoiceSig)

res=invoice_solver(invoice="items are  pen= $10, notebook= $50, bag= $200")

print(res.total)

2025/09/16 12:10:40 ERROR dspy.predict.program_of_thought: Error in code execution: Invalid Python syntax. message: 


260


# OPTIMIZERS

1. BootstrapFewShot

Learns a better prompt from a small dataset of examples.

Automatically builds few-shot prompts.

📌 Example: Customer Support Bot

In [ ]:
import dspy

# Define signature
class SupportSig(dspy.Signature):
    query = dspy.InputField()
    response = dspy.OutputField()

# Base model
support_bot = dspy.Predict(SupportSig)

# Tiny training set
trainset = [
    {"query": "How do I reset my password?", 
     "response": "Click 'Forgot Password' on login page."},
    {"query": "Where can I find my invoice?", 
     "response": "Invoices are in the Billing section of your account."},
]

# Optimizer
opt = dspy.BootstrapFewShot()

# Train (this is the key line)
optimized_bot = opt.fit(support_bot, trainset=trainset)

# Test
print(optimized_bot(query="How do I change my email address?").response)


# 2. LabeledFewShot

Like Bootstrap, but you provide explicitly labeled training data.

Ensures controlled optimization.

📌 Example: Sentiment Analysis

In [ ]:
class SentimentSig(dspy.Signature):
    text = dspy.InputField()
    sentiment = dspy.OutputField()

sentiment_model = dspy.Predict(SentimentSig)

trainset = [
    {"text": "I love this product", "sentiment": "Positive"},
    {"text": "This is terrible", "sentiment": "Negative"},
]

optimizer = dspy.LabeledFewShot()
optimized_model = optimizer(sentiment_model, trainset=trainset)

print(optimized_model(text="Not bad at all").sentiment)


# GEPA (Gradient-based Example and Prompt Augmentation)

Advanced optimizer:

Adds examples automatically.

Optimizes instructions based on metrics (accuracy, F1, BLEU, etc).

Iterative improvement loop, similar to gradient descent but for prompts.

📌 Example: Named Entity Recognition (NER)

In [ ]:
class ExtractSig(dspy.Signature):
    text = dspy.InputField()
    company = dspy.OutputField()

ner_model = dspy.Predict(ExtractSig)

trainset = [
    {"text": "Apple launched a new iPhone.", "company": "Apple"},
    {"text": "Google released AI features.", "company": "Google"},
]

optimizer = dspy.GEPA(metric="f1")
optimized_model = optimizer(ner_model, trainset=trainset)

print(optimized_model(text="Microsoft acquired GitHub").company)


# 🚀 Pipelines in DSPy 
🔹 What are Pipelines?

Think of pipelines as workflows that connect multiple DSPy modules together.

Each step in the pipeline transforms inputs → outputs, and the outputs of one module can be passed as inputs to the next.

This is where DSPy really shines, because instead of writing long monolithic prompts, you compose small reusable steps.

👉 Analogy:
Pipelines are like machine learning pipelines (Scikit-learn Pipelines or Airflow DAGs), but for LLM-based reasoning.

In [25]:
import dspy

# Define signatures
class RetrieveSignature(dspy.Signature):
    query = dspy.InputField(desc="User query")
    context = dspy.OutputField(desc="Relevant documents")

class AnswerSignature(dspy.Signature):
    question = dspy.InputField(desc="User question")
    context = dspy.InputField(desc="Documents for answering")
    answer = dspy.OutputField(desc="Final answer")

# Define modules
retriever = dspy.Predict(RetrieveSignature)
answerer = dspy.ChainOfThought(AnswerSignature)

# Define pipeline
def rag_pipeline(question):
    docs = retriever(query=question).context
    return answerer(question=question, context=docs).answer

# Example run
print(rag_pipeline("What are the applications of quantum computing?"))


Quantum computing has a wide range of potential applications, including:
*   **Drug Discovery and Materials Science:** Simulating molecular structures and chemical reactions to accelerate the development of new drugs, catalysts, and advanced materials.
*   **Financial Modeling:** Optimizing financial models for portfolio management, risk analysis, fraud detection, and high-frequency trading.
*   **Cryptography and Cybersecurity:** Breaking existing public-key encryption algorithms and enabling the development of new quantum-safe cryptographic methods and quantum key distribution.
*   **Optimization Problems:** Solving complex optimization challenges in logistics, supply chain management, traffic flow, scheduling, and manufacturing.
*   **Artificial Intelligence and Machine Learning:** Enhancing machine learning algorithms for pattern recognition, data classification, anomaly detection, and natural language processing.
*   **Data Science and Big Data Analysis:** Accelerating the process

# Example 2: Industry-level Fraud Detection (Banking)

Banks often use LLMs for fraud report triaging:

Step 1: Extract key info from the complaint.

Step 2: Classify severity (high, medium, low).

Step 3: Generate a report summary

In [26]:
class ExtractSignature(dspy.Signature):
    complaint = dspy.InputField()
    info = dspy.OutputField()

class ClassifySignature(dspy.Signature):
    info = dspy.InputField()
    severity = dspy.OutputField()

class SummarizeSignature(dspy.Signature):
    info = dspy.InputField()
    severity = dspy.InputField()
    summary = dspy.OutputField()

extractor = dspy.ChainOfThought(ExtractSignature)
classifier = dspy.Predict(ClassifySignature)
summarizer = dspy.Predict(SummarizeSignature)

def fraud_pipeline(complaint):
    info = extractor(complaint=complaint).info
    severity = classifier(info=info).severity
    summary = summarizer(info=info, severity=severity).summary
    return summary

# Example
complaint_text = "I noticed two suspicious transactions of $500 each on my account."
print(fraud_pipeline(complaint_text))


Immediately contact your bank to report the suspicious $500 transactions, providing exact dates and times, for investigation, potential charge reversal, and advice on card security.


# Multiple chain of thaught
What is Multi-Chain-of-Thought (one-line)

Run many different chain-of-thought reasoning runs for the same question (different samples / prompts / strategies), then compare/aggregate their final answers to get a more reliable result — i.e., “have many students solve the same problem and take the majority/consensus.”

In [27]:
import dspy
from collections import Counter

# 1) Signature + CoT
class QASig(dspy.Signature):
    question = dspy.InputField()
    context  = dspy.InputField()
    reasoning = dspy.OutputField()
    final_answer = dspy.OutputField()

cot = dspy.ChainOfThought(QASig)

def self_consistent_answer(question, context, tries=7):
    answers = []
    for _ in range(tries):
        # set sampling on the LM (depends on your dspy LM config)
        dspy.settings.configure(lm_kwargs={"temperature": 0.8})
        res = cot(question=question, context=context)
        ans = getattr(res, "final_answer", None) or res.get("final_answer")
        answers.append(ans.strip())
    # majority vote
    most_common, count = Counter(answers).most_common(1)[0]
    return {"answer": most_common, "raw_votes": dict(Counter(answers))}

# Usage
out = self_consistent_answer("If 3 boxes have 2,3,4 items, how many total?", "")
print(out)


{'answer': '9', 'raw_votes': {'9': 7}}


# DSPy MultiChainComparison & Evaluation (Observability)

This is where DSPy moves beyond just “making multiple reasoning paths” → it helps you compare, evaluate, and choose the best output in a structured way.

🔹 1. What is MultiChainComparison?

It’s a DSPy helper/module that allows you to run multiple chains (e.g., several CoTs, or CoT + PoT + ReAct) in parallel, then compare their outputs.

Instead of manually writing aggregation code (majority vote, semantic similarity, verifier LLM), you can plug into DSPy’s evaluation and optimization loop.

👉 Analogy: If MultiChainOfThought is like running 5 students’ answers, MultiChainComparison is like grading them and picking the best one automatically.

🔹 2. Why it matters (real-world pain point)

In production, you can’t afford to just trust one chain-of-thought.

You also can’t manually inspect every reasoning trace.

You need an automatic evaluator that can:

Check correctness.

Check factual grounding.

Check format compliance.

Choose the best candidate.

This is exactly what MultiChainComparison helps automate.

🔹 3. Example: Using MultiChainComparison

Imagine you’re building a math tutor bot. You want the model to reason step by step, but sometimes it makes arithmetic slips.

import dspy

class MathSig(dspy.Signature):
    question = dspy.InputField(desc="Math problem to solve")
    reasoning = dspy.OutputField(desc="Step-by-step reasoning")
    answer = dspy.OutputField(desc="Final numeric answer")

# Define reasoning modules
cot = dspy.ChainOfThought(MathSig)
pot = dspy.ProgramOfThought(MathSig)

# MultiChainComparison
multi = dspy.MultiChainComparison(
    modules=[cot, pot], 
    scoring="exact_match"   # or "semantic_f1", or a custom evaluator
)

res = multi(question="If a train moves at 60 km/h for 2 hours, how far?")
print(res.answer)


✅ DSPy will:

Run both CoT and PoT.

Score their answers.

Pick the best.

Return the result.

🔹 4. Industry-level Example: Customer Support Bot

Pipeline:

Retriever: Pulls knowledge base articles.

Answer Modules:

CoT (natural reasoning).

ReAct (tool-augmented with DB lookups).

PoT (structured JSON output).

MultiChainComparison: Evaluate all answers with a custom scoring function (e.g., Semantic F1 vs. ground truth answers, or LLM verifier checking for hallucinations).

def custom_score(candidate, reference):
    # Use semantic similarity, regex checks, or LLM verification
    if "account" in candidate.lower():
        return 1.0
    return 0.5

multi = dspy.MultiChainComparison(
    modules=[cot, react, pot],
    scoring=custom_score
)

out = multi(question="How do I reset my account password?",
            context="Password reset steps: ...")
print(out.answer)


✅ Benefit:

Ensures the most relevant, grounded answer gets chosen.

Reduces hallucination in sensitive industries like finance, healthcare, or enterprise SaaS.

🔹 5. Production Best Practices

Evaluator choices:

Use semantic-F1 for free-text answers.

Use exact-match for short/numeric answers.

Use a Verifier LLM for complex compliance checks.

Logging: Always log all candidate answers + scores (for observability + auditing).

Cost control: Only run MultiChainComparison for high-risk / high-value queries (gating).

Integration: Pair it with DSPy Optimizers — they can tune entire pipelines based on evaluation metrics.



“MultiChainComparison is like having multiple students solve the same question and then the teacher picks the best solution. It improves accuracy by comparing different reasoning attempts automatically.”



“In production AI, we don’t just want one answer — we want the best possible answer. MultiChainComparison lets us run multiple reasoning strategies (CoT, ReAct, PoT) and automatically score their outputs against metrics like semantic similarity or compliance rules. This gives us more reliable results, reduces hallucination, and makes pipelines auditable. Companies in finance, healthcare, and support use this approach to meet enterprise reliability standards.”



In [30]:
import dspy

# --- Signature ---
class MathSig(dspy.Signature):
    question = dspy.InputField(desc="Math problem to solve")
    reasoning = dspy.OutputField(desc="Step-by-step reasoning")
    answer = dspy.OutputField(desc="Final numeric answer")

# --- Create modules (same as before) ---
cot = dspy.ChainOfThought(MathSig)
pot = dspy.ProgramOfThought(MathSig)

# --- Simple runner + exact-match comparator ---
def run_and_compare(modules, signature, question):
    results = []
    for m in modules:
        # call module just like you would: each returns an object with .reasoning and .answer
        out = m(question=question)
        # normalize answer for exact match (strip + lowercase)
        ans_norm = getattr(out, "answer", "").strip().lower()
        results.append({"module": m, "out": out, "ans": ans_norm})
    # exact-match scoring: count matches for each answer
    counts = {}
    for r in results:
        counts[r["ans"]] = counts.get(r["ans"], 0) + 1
    # pick most common normalized answer (tie -> first module's answer)
    best_ans_norm = max(counts.items(), key=lambda x: (x[1], x[0]))[0]
    # find the result object that produced that normalized answer (prefer earlier module)
    chosen = next(r for r in results if r["ans"] == best_ans_norm)
    return chosen["out"]

# --- Run ---
out = run_and_compare([cot, pot], MathSig, question="If a train moves at 60 km/h for 2 hours, how far?")
print("Reasoning:", getattr(out, "reasoning", None))
print("Answer:", getattr(out, "answer", None))


Reasoning: To find the distance, we use the formula: Distance = Speed × Time.
Given:
Speed = 60 km/h
Time = 2 hours

Distance = 60 km/h * 2 hours
Distance = 120 km
Answer: 120


# DSPy Evaluation & Metrics
🔹 1. Why evaluation matters

LLMs are non-deterministic → same input can yield different outputs.

Without structured evaluation, you can’t tell if your AI pipeline is actually improving.

In production (banking, healthcare, legal, support), you must prove reliability with metrics.

👉 DSPy builds evaluation directly into the framework: you can attach metrics, validators, and optimizers to pipelines.

🔹 2. What kind of metrics are used?

Exact Match (EM)

Good for short answers, multiple-choice, or yes/no questions.

Example: Q: “Capital of France?” A: “Paris” → EM = 1.

Semantic F1 (token-level overlap)

Better for free-text where wording differs.

Example: “The train traveled 120 km” vs “It went 120 kilometers” → EM = 0, but F1 ~ 1.

Rouge / BLEU / Similarity scores

For summarization, translation, or long text.

LLM-as-a-Judge (verifier model)

Ask a small LM: “Which answer best matches the context?”

Used when outputs are too open-ended for EM/F1.

Custom checks / business rules

Regex, numeric bounds, JSON schema validation, compliance rules.

Example: In healthcare → answer must always include a disclaimer.

In [33]:
import dspy
from dspy.evaluate import Evaluate

# Define signature + module
class QnASignature(dspy.Signature):
    question = dspy.InputField()
    context = dspy.InputField()
    answer = dspy.OutputField()

qa_model = dspy.ChainOfThought(QnASignature)

# Define evaluation metric
def semantic_f1(pred, gold):
    pred_tokens = set(pred.lower().split())
    gold_tokens = set(gold.lower().split())
    overlap = len(pred_tokens & gold_tokens)
    precision = overlap / len(pred_tokens) if pred_tokens else 0
    recall = overlap / len(gold_tokens) if gold_tokens else 0
    if precision + recall == 0:
        return 0
    return 2 * precision * recall / (precision + recall)

# Sample dataset
dataset = [
    {"question": "What is the capital of France?", 
     "context": "France is a country in Europe. Its capital is Paris.", 
     "gold": "Paris"},
    {"question": "2+2?", "context": "Basic math", "gold": "4"}
]

# Initialize evaluator (requires devset)
evaluator = Evaluate(
    model=qa_model,
    metric=semantic_f1,
    devset=dataset,              # <-- fix: pass dataset here
    input_keys=["question", "context"],
    output_key="answer"
)

# Run evaluation
results = evaluator(program=qa_model)
print(results)


2025/09/17 01:11:45 ERROR dspy.utils.parallelizer: Error for {'question': 'What is the capital of France?', 'context': 'France is a country in Europe. Its capital is Paris.', 'gold': 'Paris'}: 'dict' object has no attribute 'inputs'. Set `provide_traceback=True` for traceback.
2025/09/17 01:11:45 ERROR dspy.utils.parallelizer: Error for {'question': '2+2?', 'context': 'Basic math', 'gold': '4'}: 'dict' object has no attribute 'inputs'. Set `provide_traceback=True` for traceback.
2025/09/17 01:11:47 INFO dspy.evaluate.evaluate: Average Metric: 0.0 / 2 (0.0%)


EvaluationResult(score=0.0, results=<list of 2 results>)
